eval.go

```diff
@@ -568,30 +568,32 @@ func (d *indexData) gatherBranches(docID uint32, mt matchTree, known map[matchTr
 }
 
 func (d *indexData) List(ctx context.Context, q query.Q, opts *ListOptions) (rl *RepoList, err error) {
-	var include func(rle *RepoListEntry) (bool, error)
+	var include func(rle *RepoListEntry) bool
 
 	q = d.simplify(q)
 	if c, ok := q.(*query.Const); ok {
 		if !c.Value {
 			return &RepoList{}, nil
 		}
-		include = func(rle *RepoListEntry) (bool, error) {
-			return true, nil
+		include = func(rle *RepoListEntry) bool {
+			return true
 		}
 	} else {
-		// We need to run a search per repo to decide if it is included.
-		include = func(rle *RepoListEntry) (bool, error) {
-			qOneRepo := query.NewAnd(
-				query.NewRepoSet(rle.Repository.Name),
-				q)
-			sr, err := d.Search(ctx, qOneRepo, &SearchOptions{
-				ShardMaxMatchCount: 1,
-				TotalMaxMatchCount: 1,
-			})
-			if err != nil {
-				return false, err
-			}
-			return len(sr.Files) > 0, nil
+		sr, err := d.Search(ctx, q, &SearchOptions{
+			ShardRepoMaxMatchCount: 1,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		foundRepos := make(map[uint32]struct{}, len(sr.Files))
+		for _, file := range sr.Files {
```

Comment:
instead of creating this intermediate map, can't we somehow go from sr.Files entry straight to the repo metadata?

Reply:
AFAICT, not without changing behavior. 

[`RepoListEntry`](https://sourcegraph.com/github.com/sourcegraph/zoekt@4a6dab85587c1e7a1719a71454bbd2d0aa7c288a/-/blob/api.go?L504-508) has fields on it (`Stats`, `IndexMetadata`) that are not available on the file match. 

If `d.repoListEntry` is reliably sorted, we could sort the file matches by the same key and do a linear merge without any additional allocations, but I couldn't find any information on whether `d.repoListEntry` is sorted.

Reply:
yeah relying on sorted order is dangerous. On more thought this code path is fine, given it only runs when the List query isn't the const true. LGTM